src/Polysemy/Internal/Scoped.hs

```diff
@@ -21,30 +22,138 @@ import Polysemy.Internal.Combinators (transform)
 --
 -- For a longer exposition, see <https://www.tweag.io/blog/2022-01-05-polysemy-scoped/>.
 --
--- Resource creation is performed by the function passed to 'Polysemy.Scoped.runScoped'.
+-- Resource allocation is performed by a function passed to 'Polysemy.Scoped.interpretScoped'.
 --
 -- The constructors are not intended to be used directly; the smart constructor 'scoped' is used like a local
 -- interpreter for @effect@.
-data Scoped (resource :: Type) (effect :: Effect) :: Effect where
-  Run :: ∀ resource effect m a . resource -> effect m a -> Scoped resource effect m a
-  InScope :: ∀ resource effect m a . (resource -> m a) -> Scoped resource effect m a
+-- 'scoped' takes an argument of type @param@, which will be passed through to the interpreter, to be used by the
+-- resource allocation function.
+--
+-- As an example, imagine an effect for writing lines to a file:
+--
+-- > data Write :: Effect where Write :: Text -> Write m ()
+-- > makeSem ''Write
+--
+-- If we now have the following requirements:
+--
+-- 1. The file should be opened and closed right before and after the part of the program in which we write lines
+-- 2. The file name should be specifiable at the point in the program where writing begins
+-- 3. We don't want to commit to IO, lines should be stored in memory when running tests
+--
+-- Then we can take advantage of 'Scoped' to write this program:
+--
+-- > prog :: Member (Scoped FilePath resource Write) r => Sem r ()
+-- > prog = do
+-- >   scoped "file1.txt" do
+-- >     write "line 1"
+-- >     write "line 2"
+-- >   scoped "file2.txt" do
+-- >     write "line 1"
+-- >     write "line 2"
+--
+-- Here 'scoped' creates a prompt for an interpreter to start allocating a resource for @"file1.txt"@ and handling
+-- @Write@ actions using that resource.
+-- When the 'scoped' block ends, the resource should be freed.
+--
+-- The interpreter may look like this:
+--
+-- > interpretWriteFile :: Member (Embed IO) => InterpreterFor (Scoped FilePath Handle Write) r
+-- > interpretWriteFile =
+-- >   interpretScoped (\ name use -> bracket (openFile name WriteMode) hClose use) \ handle -> \case
+-- >     Write line -> embed (Text.hPutStrLn handle line)
+--
+-- Essentially, the @bracket@ is executed at the point where @scoped@ was called, wrapping the following block.
+-- When the second @scoped@ is executed, another call to @bracket@ is performed.
+--
+-- The effect of this is that the operation that uses @Embed IO@ was moved from the call site to the interpreter, while
+-- the interpreter may be executed at the outermost layer of the app.
+--
+-- This makes it possible to use an in-memory interpreter for testing:
+--
+-- > interpretWriteMap :: Member (Output (FilePath, Text)) r => InterpreterFor (Scoped FilePath FilePath Write) r
+-- > interpretWriteMap =
+-- >   interpretScoped (\ name use -> use name) \ name -> \case
```

Comment:
I'm not exactly sure how `use name` type checks here. In order to do this, we need `parameter ~ resource`, right? Do you *need* that to be true in order to write this program? I guess so that you can `output (name, line)` and have it type check as well?

Reply:
well yeah, it's a degenerate case. we're simply not doing anything in the allocator.